### PR TITLE
Refactor second batch of unary elementwise ops of MHLO unittests

### PR DESCRIPTION
**Pull request by heidebro**

### CODE REVIEW

**Review by henri-gruender — COMMENTED** [review]
There is some inconsistent use of float and double literals. Also I would suggest testing some negative values for `round`.

**Review by henri-gruender — DISMISSED** [review]

File: reference-implementation/unittests/mhlo.cpp [hunk 1]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+  }
+  {
+    Tensor1D<float, 2> x{M_Ef32 * M_Ef32, M_Ef32 * M_Ef32 * M_Ef32};
+    Tensor1D<float, 2> expected_result{2.126928f, 3.048587f};
+    Tensor1D<float, 2> result = mhlo::log_plus_one(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
+  }
+  {
+    Tensor2D<double, 2, 2> x{0.0f, 1.0f, 2.0f, 3.0f};
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:437]:
```suggestion
    Tensor2D<double, 2, 2> x{0.0, 1.0, 2.0, 3.0};
```

File: reference-implementation/unittests/mhlo.cpp [hunk 2]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+    Tensor2D<double, 2, 2> expected_result{0.0f, 0.693147f, 1.098612f,
+                                           1.386294f};
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:439]:
```suggestion
    Tensor2D<double, 2, 2> expected_result{0.0, 0.693147, 1.098612,
                                           1.386294};
```

File: reference-implementation/unittests/mhlo.cpp [hunk 3]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+    Tensor2D<double, 2, 2> result = mhlo::log_plus_one(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:442]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 4]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+  }
+  {
+    Tensor3D<double, 2, 2, 1> x{0.0f, 1.0f, 2.0f, 3.0f};
+    Tensor3D<double, 2, 2, 1> expected_result{0.0f, 0.693147f, 1.098612f,
+                                              1.386294f};
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:447]:
```suggestion
    Tensor3D<double, 2, 2, 1> x{0.0, 1.0, 2.0, 3.0};
    Tensor3D<double, 2, 2, 1> expected_result{0.0, 0.693147, 1.098612,
                                              1.386294};
```

File: reference-implementation/unittests/mhlo.cpp [hunk 5]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+    Tensor3D<double, 2, 2, 1> result = mhlo::log_plus_one(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:450]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 6]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+  }
+  {
+    Tensor4D<double, 2, 2, 1, 2> x{0.0f, 1.0f, 2.0f, 3.0f,
+                                   4.0f, 5.0f, 6.0f, 7.0f};
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:454]:
```suggestion
    Tensor4D<double, 2, 2, 1, 2> x{0.0, 1.0, 2.0, 3.0,
                                   4.0, 5.0, 6.0, 7.0};
```

File: reference-implementation/unittests/mhlo.cpp [hunk 7]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+    Tensor4D<double, 2, 2, 1, 2> expected_result{
+        0.0f,      0.693147f, 1.098612f, 1.386294f,
+        1.609437f, 1.791759f, 1.945910f, 2.079441f};
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:457]:
```suggestion
        0.0,      0.693147, 1.098612, 1.386294,
        1.609437, 1.791759, 1.945910, 2.079441};
```

File: reference-implementation/unittests/mhlo.cpp [hunk 8]
```diff
@@ -419,17 +419,46 @@ TEST(mhlo, log) {
...
+    Tensor4D<double, 2, 2, 1, 2> result = mhlo::log_plus_one(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:460]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 9]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+    Tensor1D<float, 2> result = mhlo::sin(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
+  }
+  {
+    Tensor2D<double, 2, 2> x{2 * M_PIf32, 0.0, -0.5, 0.5};
+    Tensor2D<double, 2, 2> expected_result{0.0, 0.0, -0.479426, 0.479426};
+    Tensor2D<double, 2, 2> result = mhlo::sin(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:574]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 10]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+  }
+  {
+    Tensor3D<double, 1, 2, 2> x{2 * M_PIf32, 0.0, -0.5, 0.5};
+    Tensor3D<double, 1, 2, 2> expected_result{0.0, 0.0, -0.479426, 0.479426};
+    Tensor3D<double, 1, 2, 2> result = mhlo::sin(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:581]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 11]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+  }
+  {
+    Tensor4D<double, 2, 1, 2, 1> x{2 * M_PIf32, 0.0, -0.5, 0.5};
+    Tensor4D<double, 2, 1, 2, 1> expected_result{0.0, 0.0, -0.479426, 0.479426};
+    Tensor4D<double, 2, 1, 2, 1> result = mhlo::sin(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:588]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 12]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+    Tensor1D<float, 2> result = mhlo::sqrt(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
+  }
+  {
+    Tensor2D<double, 2, 2> x{2.0, 3.0, 10.0, 1.0};
+    Tensor2D<double, 2, 2> expected_result{1.414213, 1.732050, 3.162277, 1.0};
+    Tensor2D<double, 2, 2> result = mhlo::sqrt(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:614]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 13]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+  }
+  {
+    Tensor4D<double, 2, 2, 1, 2> x{2.0, 3.0, 10.0, 1.0, 18.0, 9.0, 5.0, 25.0};
+    Tensor4D<double, 2, 2, 1, 2> expected_result{
+        1.414213, 1.732050, 3.162277, 1.0, 4.242640, 3.0, 2.236067, 5.0};
+    Tensor4D<double, 2, 2, 1, 2> result = mhlo::sqrt(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:630]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 14]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+    Tensor1D<float, 2> result = mhlo::tanh(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
+  }
+  {
+    Tensor2D<double, 2, 2> x{0.0, 1.0, -1.0, 0.0};
+    Tensor2D<double, 2, 2> expected_result{0.0, 0.761594, -0.761594, 0.0};
+    Tensor2D<double, 2, 2> result = mhlo::tanh(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:656]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

File: reference-implementation/unittests/mhlo.cpp [hunk 15]
```diff
@@ -483,78 +512,164 @@ TEST(mhlo, round) {
...
+  }
+  {
+    Tensor4D<double, 3, 1, 1, 2> x{0.0, 1.0, -1.0, 0.0, M_PIf64, -M_2_PIf64};
+    Tensor4D<double, 3, 1, 1, 2> expected_result{0.0, 0.761594, -0.761594,
+                                                 0.0, 0.996272, -0.562593};
+    Tensor4D<double, 3, 1, 1, 2> result = mhlo::tanh(x);
+
+    EXPECT_THAT(result, Pointwise(FloatNear(EPSILON), expected_result));
```
Comment (marbre) [inline reference-implementation/unittests/mhlo.cpp:671]:
```suggestion
    EXPECT_THAT(result, Pointwise(DoubleNear(EPSILON), expected_result));
```

**Review by marbre — CHANGES_REQUESTED** [review]
There are some more tensors with double type that get initialized with floats. In addition, we should probably use `DoubleNear` instead of `FloatNear` when comparing doubles, though haven't tested it myself.

**Review by marbre — APPROVED** [review]
